### ISSUE TITLE
AOT support 

### ISSUE THREAD

**Issue by stevefan1999-personal** [zherczeg/sljit#228]
I wonder if we can use the emitted code statically, so we can save the JIT code into an object file for caching (for example). This means the codegen should not emit any position-dependent code (I think it should be PIC by default) and do not reference any runtime-dependent address as well. Is it possible for SLJIT to do this? 

(Out of context but as a query) Far as I can see the stackless in SLJIT means there aren't any push/pop operation directly on the stack pointer, and you have to maintain an external stack created on a heap right? How does that translate in terms of assembly?

**Comment by zherczeg** [comment]
All code is PIC except branches / calls.
- Calls to other pic code must use indirect calls. This is slower than normal calls.
- The code must not be too big. This is especially complicated on MIPS. Branches only have 16 bit offset, so +-128K is the maximum range. There is a `J` instruction, but the target address must be in the same 256MB region, which is not exactly PIC friendly. For larger branches, the absolute address is encoded in the instruction stream.

You can allocate a stack area for local data for every function. You need to manage your variables manually there (hence the sljit name comes). See the last argument of `sljit_emit_enter`. The `SLJIT_MEM1(SLJIT_SP)` is the base address of the local data.

**Comment by nacho00112** [comment]
So I actually can use the stack but have to manage its usage manually?
I've confused with this theme.

**Comment by zherczeg** [comment]
Yes. `SLJIT_MEM1(SLJIT_SP), offset` is the way to read/write the stack. The stack is just a linear memory area.

**Comment by nacho00112** [comment]
This guy seems to have made a stackful version of the project
https://github.com/GregorR/sfjit
what do you think about it?

**Comment by zherczeg** [comment]
It seems the author added alloca support, but still no local management. Btw building a higher level generic compiler based on sljit is possible. It can manage locals. E.g. the WebAssembly compilers based on sljit do this.
https://github.com/partic2/pwart
https://github.com/Samsung/walrus

**Comment by zherczeg** [comment]
Back to the topic of AOT, I have an idea. A serialize/de-serialize interface could save/restore the internal representation of the compiler. The loaded code could be compiled with generate code. This is not as fast as loading PIC, but the compiler can optimize branches, and still much faster than a full compilation.

**Comment by zherczeg** [comment]
I have added serialization to the code:
https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitLir.h#L2288

Since this is a new feature, there might be bugs in the code. With this interface the majority of the compilation can be skipped without giving up performance.

**Comment by stevefan1999-personal** [comment]
> I have added serialization to the code: https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitLir.h#L2288
> 
> Since this is a new feature, there might be bugs in the code. With this interface the majority of the compilation can be skipped without giving up performance.

Is that also backward/forward compatible? I might consider saving the AOT results and cache to local filesystem (similar to GAC)

**Comment by zherczeg** [comment]
It has a version number, so you can detect if your cache is not valid anymore, and can recompile the code:
https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitSerialize.c#L46

**Comment by jbsolomon** [comment]
Sorry I'm a bit confused @zherczeg but line 2288 no longer refers to anything having to do with serialization.  I was thinking of using this library to target 64-bit RISC-V for runtime use targeting a riscv vm (using libriscv).  I need to be able to disable JIT execution for hosts that disallow it (such as ios).  I'll need to locate the assembled binary into the guest's virtual address space.  Is this library an option for doing that?  I couldn't quite get clear on how much JIT execution itself is built into the design of sljit as opposed to being orthogonal to the generation of the executable code, and location of the assembly.  A bit new to this!

**Comment by zherczeg** [comment]
Just search for the `Serialization functions` part in `sljitLir.h`.

I don't understand the question. AOT in sljit is a partial compilation, where majority of the compilation is done early, and only the final code generation steps are done before execution. This final step needs a few API calls, so if you don't want to use JIT just don't execute these calls.
